core/src/audio_subsystem.cpp

```diff
@@ -0,0 +1,206 @@
+}
+
+void AudioSubsystem::stop(components::AudioSource& source) {
+    if (!source.get_channel())
+        return;
+
+    source.set_playing(false);
+    source.get_channel()->setPaused(true);
+    source.set_channel(nullptr);
+}
+
+void AudioSubsystem::stop() {
+    auto sceneOpt = Scene::get_active_scene();
+    if (!sceneOpt)
+        return;
+
+    auto scene = sceneOpt.value();
+    auto view = scene.get().get_registry().view<components::AudioSource>();
+
+    for (entt::entity entt : view) {
+        auto goOpt = scene.get().get_game_object_from_handle(entt);
+        if (!goOpt)
+            continue;
+
+        auto go = goOpt.value();
+        components::AudioSource& source = go.get_component<components::AudioSource>();
+
+        stop(source);
+    }
+}
+
+void AudioSubsystem::update() {
+    unsigned int pos = 0;
+    unsigned int len_ms = 0;
+    bool playing = false;
+    bool paused = false;
+
+    m_result = m_system->update();
+    ErrorCheck(m_result);
+
+    auto sceneOpt = Scene::get_active_scene();
+    if (!sceneOpt)
+        return;
+
+    auto scene = sceneOpt.value();
+    auto view = scene.get().get_registry().view<components::AudioSource>();
+
+    for (entt::entity entt : view) {
+        auto goOpt = scene.get().get_game_object_from_handle(entt);
+        if (!goOpt)
+            continue;
+
+        auto go = goOpt.value();
+        components::AudioSource& source = go.get_component<components::AudioSource>();
+
+        if (source.get_channel()) {
+            FMOD::Sound* current_sound = nullptr;
+            m_result = source.get_channel()->isPlaying(&playing);
+
+            if (m_result == FMOD_ERR_INVALID_HANDLE) {
+                source.set_channel(nullptr);
+                continue;
+            }
+
+            m_result = source.get_channel()->getPaused(&paused);
+            ErrorCheck(m_result);
+            m_result = source.get_channel()->getPosition(&pos, FMOD_TIMEUNIT_MS);
+            ErrorCheck(m_result);
+            source.get_channel()->getCurrentSound(&current_sound);
+
+            if (current_sound) {
+                m_result = current_sound->getLength(&len_ms, FMOD_TIMEUNIT_MS);
+                ErrorCheck(m_result);
+            }
+        }
+    }
+}
+
+void AudioSubsystem::set_loop(components::AudioSource& source, bool loops) {
+    source.set_loops(loops);
+
+    m_result = source.get_sound()->setMode(source.get_loops() ? FMOD_LOOP_NORMAL : FMOD_LOOP_OFF);
+    ErrorCheck(m_result);
+}
+
+void AudioSubsystem::add_sound(components::AudioSource& source) {
+    log::info("TRYING TO LOAD AUDIO FILE {}", source.get_path().string());
+    m_result =
+        m_system->createSound(source.get_path().string().c_str(), FMOD_3D_LINEARROLLOFF, nullptr, &source.get_sound());
+    ErrorCheck(m_result);
+
+    m_result = source.get_sound()->set3DMinMaxDistance(0.0f, 50000.0f);
+    ErrorCheck(m_result);
+    set_loop(source, source.get_loops());
+    source.get_sound()->setDefaults(12000, 128);
```

Comment:
Please put these numbers inside their own `constexpr` to avoid having magic numbers